[Typechecker] Allow static functions to overload enum cases

theblixguy opened the pull request:
Fixes an issue with redeclaration checking where a static function would incorrectly get marked as a redeclaration of an enum element declaration, even if their types were different. See https://twitter.com/an0/status/1105187953834369025

Resolves [SR-10084](https://bugs.swift.org/browse/SR-10084).

theblixguy commented:
cc @slavapestov @xedin

Commit:
[typechecker] fix an issue with redeclaration checking

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -2090,6 +2090,14 @@ bool swift::conflicting(ASTContext &ctx,
+  }
+
   // Functions always conflict with non-functions with the same signature.
   // In practice, this only applies for zero argument functions.
```
I think this comment is misleading -- this is the case we're hitting here, and its not a zero argument function.

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -2090,6 +2090,14 @@ bool swift::conflicting(ASTContext &ctx,
 
+  // Static functions do not conflict with enum element decls if the types
+  // are different
+  if (sig1.IsStaticFunction == sig2.IsEnumElement) {
```
Can we fold this into the following 'if'? I think even non-static functions should not conflict with enum elements.

In fact, is there any harm to treating enum elements as _functions_ for purposes of disambiguation?

hamishknight replied:
I agree, I think we should treat functions and enum elements the same for the purposes of overload signature checking. How about renaming `IsFunction` to something like `IsFunctionLike`, and then initialising it with `isa<AbstractFunctionDecl>(this) || isa<EnumElementDecl>(this)`?

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -2090,6 +2090,14 @@ bool swift::conflicting(ASTContext &ctx,
+
   // Functions always conflict with non-functions with the same signature.
   // In practice, this only applies for zero argument functions.
   if (sig1.IsFunction != sig2.IsFunction)
```
To elaborate on the first point, I think we might want to just check IsEnumElement here.

slavapestov commented on test/decl/overload.swift:
```diff
@@ -476,5 +476,20 @@ extension SR7250 where T : P3 {
+enum SR_10084_E {
+  case foo(SR_10084_S) // expected-note {{'foo' previously declared here}}
+    
+  static func foo(_ name: String) -> SR_10084_E { // Ok
```
There should be another test case for the opposite order too, where the static function comes first

slavapestov reviewed (commented):
Thanks!

hamishknight commented on lib/AST/Decl.cpp:
```diff
@@ -2293,6 +2304,10 @@ CanType ValueDecl::getOverloadSignatureType() const {
 
+  if (isa<EnumElementDecl>(this)) {
+    return this->getInterfaceType()->getCanonicalType();
+  }
```
Note that by giving enum elements different overload types, the overloading of enum cases by associated type becomes legal, e.g allowing:

```swift
enum E {
  case fn(Int)
  case fn(String)
}
```

This is something we ultimately want for [SE-0155], but I don't believe the implementation is currently in place to support it. For now at least, I think the overload type variant of `swift::conflicting` should check to see if it's comparing two enum elements and always consider them to conflict to keep the above example illegal until the implementation is ready.

In addition, I think `mapSignatureFunctionType` should be used here in order to ensure we don't consider things like `@escaping`. For example, for the following:

```swift
enum E {
  case fn(() -> Void)

  static func fn(_ x: @escaping () -> Void) -> E {
    fatalError()
  }
}
```

The method `fn` has the overload type `(E) -> (() -> Void) -> E` (with the `@escaping` having been stripped away by `mapSignatureFunctionType`), whereas the enum element would have the overload type `(E) -> (@escaping () -> Void) -> E`. By using `mapSignatureFunctionType` on the enum element, it would also have its `@escaping` stripped away, bringing the signatures in line.
 


  [SE-0155]: https://github.com/apple/swift-evolution/blob/master/proposals/0155-normalize-enum-case-representation.md

slavapestov replied:
Yeah, we should leave overloading enum cases by type (and label) for a separate PR. One place where enum cases appear the mangling is for resilient case index symbols and I need to check if we handle overloading there.

theblixguy commented:
@slavapestov @hamishknight Thanks for your feedback, I have adjusted my code.

There still seems to be an issue, because I am getting lots of errors when building stdlib, for example:

> /Users/suyashsrijan/Documents/swift-src/swift/stdlib/public/core/Tuple.swift.gyb:165:13: error: invalid redeclaration of '!='
public func != <A : Equatable, B : Equatable, C : Equatable>(lhs: (A,B,C), rhs: (A,B,C)) -> Bool {
            ^
/Users/suyashsrijan/Documents/swift-src/swift/stdlib/public/core/Builtin.swift:114:15: note: '!=' previously declared here
internal func != (
              ^

theblixguy commented:
Seems to be happening because of the check `if (sig1.IsEnumElement == sig2.IsEnumElement) { return true }`

Commit:
[typechecker] use mapSignatureFunctionType in getOverloadSignatureType() to compute the type for the enum element decl, etc

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -2090,6 +2090,17 @@ bool swift::conflicting(ASTContext &ctx,
+    return true;
+  }
+
+  // A function does not conflict with an enum element if their types differ.
```
What if sig1 is an enum element and sig2 is a function? Enum cases can appear anywhere inside an enum body, including after functions.

theblixguy replied:
@slavapestov True, I have added that case now.

Commit:
[typechecker] allow matching enums to functions as well, not just functions to enums

Commit:
[typechecker] fix the check for two enums

Commit:
[typechecker] check for nominal types as well when comparing enum elements

Commit:
[test] add more tests

theblixguy commented:
I have added some more tests to ensure non-functions can't redeclare an enum element (or vice versa). We only had one test for this, but I have added some more test cases to ensure we don't regress.

Commit:
[typechecker] check for typealias as well

theblixguy commented:
@slavapestov Do you have any additional feedback or is this good to go?

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

theblixguy commented:
All tests have passed 🎉 Is this worth cherry-picking to 5.1 as well?
